src/dspeed/processors/ml.py

```diff
@@ -0,0 +1,355 @@
+    temp[0] = np.dot(x_in, kernel)
+
+    if activation_func == ord("s"):
+        sigmoid(temp, x_out)
+    elif activation_func == ord("r"):
+        relu(temp, x_out)
+    elif activation_func == ord("l"):
+        leaky_relu(temp, x_out)
+    elif activation_func == ord("m"):
+        softmax(temp, x_out)
+    elif activation_func == ord("t"):
+        x_out[0] = np.tanh(temp[0])
+
+
+@guvectorize(
+    [
+        "void(float32[::1], float32[::1], float32, char, float32[:])",
+        "void(float64[::1], float64[::1], float64, char, float64[:])",
+    ],
+    "(n),(n),(),()->()",
+    **nb_kwargs,
+)
+def classification_layer_with_bias(
+    x_in: np.ndarray,
+    kernel: np.ndarray,
+    bias: np.ndarray,
+    activation_func: np.int8,
+    x_out: float,
+) -> None:
+    """
+    this is the same as dense_layer_with_bias but the final output is a single number f(x.W+bs)
+
+    Parameters
+    ----------
+    w_in
+        the input waveform shape n.
+    kernel
+        the matrix of weights shape (n x 1).
+    bias
+        the bias in this case a single value.
+    activation_func
+        the activation function to use specify with char:
+        s - sigmoid
+        r - relu
+        l - leaky relu
+        m - softmax
+        t - tanh
+    x_out
+        the output value.
+
+    JSON Configuration Example
+    --------------------------
+
+    .. code-block :: json
+
+        "classifier": {
+            "function": "dense_layer_with_bias",
+            "module": "dspeed.processors",
+            "args": ["wf_blsub", "db.kernel", "db.bias","'s'", "classifier"],
+            "unit": "ADC"
+        }
+    """
+    x_out[0] = np.nan
+
+    if np.isnan(x_in).any():
+        return
+
+    temp = np.zeros_like(x_out)
+    temp[0] = np.dot(x_in, kernel) + bias
+
+    if activation_func == ord("s"):
+        sigmoid(temp, x_out)
+    elif activation_func == ord("r"):
+        relu(temp, x_out)
+    elif activation_func == ord("l"):
+        leaky_relu(temp, x_out)
+    elif activation_func == ord("m"):
+        softmax(temp, x_out)
+    elif activation_func == ord("t"):
+        x_out[0] = np.tanh(temp[0])
+
+
+@guvectorize(
+    [
+        "void(float32[:], float32[:], float32[:], float32[:])",
+        "void(float64[:], float64[:], float64[:], float64[:])",
+    ],
+    "(n),(n),(n)->(n)",
+    **nb_kwargs,
+)
+def normalisation_layer(
+    x_in: np.ndarray, means: np.ndarray, variances: np.ndarray, x_out: np.ndarray
+) -> None:
+    """
+    Normalisation layer, (x_in - mu)/np.sqrt(variance)
+    Note this is variance not standard deviation
+
+    Parameters
+    ----------
+    w_in
+        the input waveform shape n.
+    means
+        array of means for each input value shape n.
+    variances
+        array of variances for each input value shape n.
+    x_out
+        the output vector shape n.
+
+    JSON Configuration Example
+    --------------------------
+
+    .. code-block :: json
+
+    "wf_normed": {
```

Comment:
The doc generation is failing with " /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/dspeed/processors/ml.py:docstring of dspeed.processors.ml.normalisation_layer:21: WARNING: Definition list ends without a blank line; unexpected unindent. [docutils]". I think this code block needs an extra indent